src/test/kotlin/org/jitsi/nlj/rtp/bandwidthestimation/BandwidthEstimationTest.kt

```diff
@@ -0,0 +1,238 @@
+        executor.schedule({ receiver(packet) }, delay.toNanos(), TimeUnit.NANOSECONDS)
+    }
+}
+
+class PacketReceiver(
+    private val clock: Clock,
+    private val estimator: BandwidthEstimator,
+    private val ctx: DiagnosticContext,
+    val rateReceiver: (Bandwidth) -> Unit
+) {
+    private val timeSeriesLogger: TimeSeriesLogger = TimeSeriesLogger.getTimeSeriesLogger(this.javaClass)
+    private var seq = 0
+
+    fun receivePacket(packet: SimulatedPacket) {
+        val now = clock.instant()
+        assert(packet.sendTime <= now)
+        /* All delay is send -> receive in this simulation, so one-way delay is rtt. */
+        estimator.onRttUpdate(now, Duration.between(packet.sendTime, now))
+        estimator.processPacketArrival(now, packet.sendTime, now, seq, packet.packetSize)
+        seq++
+        val bw = estimator.getCurrentBw(now)
+        if (timeSeriesLogger.isTraceEnabled) {
+            timeSeriesLogger.trace(ctx.makeTimeSeriesPoint("bw", now).addField("bw", bw.bps))
+        }
+        rateReceiver(bw)
+    }
+}
+
+class BandwidthEstimationTest : ShouldSpec() {
+    init {
+        /* Internals of GoogleCc use ConfigurationService at construct time. */
+        LibJitsi.start()
+    }
+    private val scheduler: FakeScheduledExecutorService = spy()
+    private val clock: Clock = scheduler.clock
+
+    private val ctx = DiagnosticContext(clock)
+    init {
+        /* Emulate the fields that jitsi-videobridge puts in its DiagnosticContexts. */
+        ctx["conf_name"] = "BandwidthEstimationTest"
+        ctx["conf_creation_time_ms"] = clock.instant().toEpochMilli()
+        ctx["endpoint_id"] = "00000000"
+    }
+
+    private val logger = LoggerImpl(BandwidthEstimationTest::class.qualifiedName)
+    private val estimator: BandwidthEstimator = GoogleCcEstimator(ctx, logger)
+
+    private val rtt: Duration = Duration.ofMillis(200)
+    private val bottleneckRate = 4.mbps
+
+    private val generator: PacketGenerator = PacketGenerator(scheduler, clock, { bottleneck.enqueue(it) })
+    private val bottleneck: PacketBottleneck = PacketBottleneck(scheduler, clock, ctx) { delayer.enqueue(it) }
+    private val delayer: PacketDelayer = PacketDelayer(scheduler, rtt) { receiver.receivePacket(it) }
+    private val receiver: PacketReceiver = PacketReceiver(clock, estimator, ctx) { generator.rate = it }
+
+    init {
+        "Running bandwidth estimation test" {
+            should("work correctly") {
+                bottleneck.rate = bottleneckRate
+                generator.rate = estimator.getCurrentBw(clock.instant())
+
+                bottleneck.start()
+                generator.start()
+
+                scheduler.runUntil(clock.instant().plus(120, ChronoUnit.SECONDS))
+
+                generator.stop()
+                bottleneck.stop()
+
+                val finalBw = estimator.getCurrentBw(clock.instant())
+                finalBw.bps.shouldBeGreaterThan((bottleneckRate / 1.2).bps)
```

Comment:
i think there's a `plusOrMinus` helper in the test lib, but i have a feeling i mentioned this before and you said there was something about it which didn't work here...

Reply:
if not, it looks like there's also `shouldBeBetween`

Reply:
When bandwidth was a Float, we couldn't use `shouldBeBetween`, because it doesn't exist for Float for some reason.  Now that it's Double, we can, so I'll change that.

`plusOrMinus` isn't applicable, though, since this is a multiplicative factor not an additive one.